lib/chef/telemetry/decision/file.rb

```diff
@@ -0,0 +1,101 @@
+
+require "chef-config/windows"
+require "chef-config/path_helper"
+require "yaml"
+require "date"
+
+class Chef
+  class Telemetry
+    class Decision
+
+      # Represents a telemetry opt-in or out recorded on disk.
+      class File
+        DECISION_FILE = "telemetry_options.yaml".freeze
+
+        attr_reader :logger, :contents, :location
+        attr_accessor :local_dir # Optional local path to use to seek
+
+        def initialize(opts)
+          @opts = opts
+          @logger = opts[:logger]
+          @contents_ivar = nil
+          @location = nil
+        end
+
+        # Checks for the existence of a decision file, and returns true if the decision was to enable.
+        # Returns false if the decision was not made or if it was not enabled.
+        def opt_in?
+          read_decision_file
+          !contents.nil? && contents[:enabled]
+        end
+
+        # Checks for the existence of a decision file, and retruns true if the decision was to disable.
+        # Returns false if the decision was not made or if it was enabled.
+        def opt_out?
+          read_decision_file
+          !contents.nil? && !contents[:enabled]
+        end
+
+        # Writes a decision file to disk in the location specified,
+        # with the content given.
+        # @return Array of Errors
+        def persist(decision, dir, content = {})
+          content[:decision_time] = DateTime.now.to_s
+          content[:enabled] = decision
+          @contents = content
+
+          begin
+            msg = "Could not create directory for telemetry optin/out decision #{dir}"
+            FileUtils.mkdir_p(dir)
+            msg = "Could not write telemetry optin/out decision file #{dir}/#{DECISION_FILE}"
+            ::File.write("#{dir}/#{DECISION_FILE}", YAML.dump(content))
+            return []
+          rescue StandardError => e
+            logger.info "#{msg}\n\t#{e.message}"
+            logger.debug "#{e.backtrace.join("\n\t")}"
+            return [e]
+          end
+        end
+
+        # Returns true if a decision file exists.
+        def persisted?
+          !!seek
+        end
+
+        private
+
+        # Look for a decsion file in several locations.
+        def seek
+          return location if location
+
+          candidates = []
+
+          # Include the user home directory ~/.chef
+          candidates << ChefConfig::PathHelper.home(".chef/#{DECISION_FILE}")
+
+          # TODO: include software installation directory, eg /opt/chef
+
+          # Include /etc/chef if on unix-like
+          candidates << "/etc/chef/#{DECISION_FILE}" unless ChefConfig.windows?
```

Comment:
Can we include C:/chef for windows, for consistency of "in the install directory" on both platforms?

Reply:
Certainly, I just didn't know what path(s) to include... any others?

Reply:
I've reworked this to default to using the home directory of the user, like license-acceptance. I've left a comment for future work in which we might add in per-product directories, but that might actually turn it into a globbing operation - there are a lot of variants due to the omnibus distros.